Userland/Libraries/LibWeb/HTML/EventSource.cpp

```diff
@@ -130,8 +130,37 @@ WebIDL::ExceptionOr<JS::NonnullGCPtr<EventSource>> EventSource::construct_impl(J
         else {
             event_source->announce_the_connection();
 
-            auto process_body_chunk = JS::create_heap_function(realm.heap(), [event_source](ByteBuffer body) {
-                event_source->interpret_response(body);
+            StringBuilder pending_line = MUST(StringBuilder::create());
+
+            auto process_body_chunk = JS::create_heap_function(realm.heap(), [event_source, pending_line](ByteBuffer body) mutable {
+                StringView body_str = body;
+
+                auto last_linebreak = body_str.find_any_of("\r\n"sv, StringView::SearchDirection::Backward);
+
+                StringView remainder;
+                if (last_linebreak.has_value()) {
+                    size_t end = last_linebreak.value() + 1;
+
+                    auto new_completed = body_str.substring_view(0, end);
+
+                    if (pending_line.is_empty()) {
+                        event_source->interpret_response(new_completed);
+                    } else {
+                        MUST(pending_line.try_append(new_completed));
+
+                        event_source->interpret_response(pending_line.string_view());
+
+                        pending_line.clear();
+                    }
+
+                    remainder = body_str.substring_view(end);
+                } else {
+                    remainder = body_str;
+                }
+
+                if (!remainder.is_empty()) {
+                    MUST(pending_line.try_append(remainder));
+                }
```

Comment:
This all feels like a very verbose solution. Is the bug just that `interpret_response` will process the last line even if it does not end with a newline? If so, can we just not process that last line in `interpret_response` and hold onto the remainder there, until we receive the next chunk?

Reply:
I agree that it's verbose, but the reason I have it out here is that it's only relevant to the current connection. Keeping the leftovers in an instance member would require an extra step to clear it when reconnecting

Reply:
also, from my reading of `for_each_line`, it wouldn't do anything different for strings ending with newlines, so that needs to be checked outside anyway

Reply:
Sorry for the delay in getting back to this, had a busy week last week.

Keeping it local to the connection callback is fine, but I'd like to massage the code a bit still. Does the following work for your test case?
```diff
diff --git a/Userland/Libraries/LibWeb/HTML/EventSource.cpp b/Userland/Libraries/LibWeb/HTML/EventSource.cpp
index a0fa53edde..7f23a267bf 100644
--- a/Userland/Libraries/LibWeb/HTML/EventSource.cpp
+++ b/Userland/Libraries/LibWeb/HTML/EventSource.cpp
@@ -130,9 +130,22 @@ WebIDL::ExceptionOr<JS::NonnullGCPtr<EventSource>> EventSource::construct_impl(J
         else {
             event_source->announce_the_connection();

-            auto process_body_chunk = JS::create_heap_function(realm.heap(), [event_source](ByteBuffer body) {
-                event_source->interpret_response(body);
+            auto process_body_chunk = JS::create_heap_function(realm.heap(), [event_source, pending_data = ByteBuffer()](ByteBuffer body) mutable {
+                if (pending_data.is_empty())
+                    pending_data = move(body);
+                else
+                    pending_data.append(body);
+
+                auto last_line_break = AK::StringUtils::find_any_of(pending_data, "\r\n"sv, AK::StringUtils::SearchDirection::Backward);
+                if (!last_line_break.has_value())
+                    return;
+
+                auto end_index = *last_line_break + 1;
+                event_source->interpret_response({ pending_data.bytes().slice(0, end_index) });
+
+                pending_data = MUST(pending_data.slice(end_index, pending_data.size() - end_index));
             });
+
             auto process_end_of_body = JS::create_heap_function(realm.heap(), []() {
                 // This case is handled by `process_event_source_end_of_body` above.
             });
```

Reply:
Seems to (actually, with both versions I occasionally see it still fail, but it's significantly more reliable than before)

Should I take this code or do you want to open a new PR?

Reply:
> with both versions I occasionally see it still fail

Same here, but I'm not sure that it's an issue with the EventSource class itself - it's likely something going on with the unbuffered request handler in Fetch or our Streams implementation that passes the data up to EventSource. I've also been looking into an issue where navigating away from a page doesn't close the SSE connection.

> Should I take this code or do you want to open a new PR?

Nah, feel free to take it

Reply:
Replaced code as suggested